### PR TITLE
Rename modeling exercise server routes to follow REST guidelines

### PR DESCRIPTION
**Pull request by maxr96**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] ~Client: I added multiple screenshots/screencasts of my UI changes~
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

We want to be consistent in naming our REST Endpoints. The changed endpoints didn't follow our REST Guidelines.

### Description

The changes reflects the table of REST routes https://confluence.ase.in.tum.de/pages/viewpage.action?pageId=30016376
Changed routes in this PR:
- GET modeling-editor/{participationId} -> participations/{participationId}/latest-submission
- PUT modeling-submissions/{submissionId}/feedback -> modelling-submissions/{submissionId}/assessment
- PUT exercise/{exerciseId}/submission/{submissionId}/modelingExampleAssessment -> modeling-submissions/{submissionId}/example-assessment
- GET modeling-submissions/{submissionId}/result -> modeling-submissions/{submissionId}/assessment
- PUT modeling-exercises -> modeling-exercises/{exerciseId}

**Test Coverage**
Thanks to @madwau all the REST endpoints (Modeling assessment, submission and exercise)  are covered with tests, I adjusted the tests to target renamed endpoints and further improved the test coverage.
Screenshot_2020-03-05 de tum in www1 artemis web rest
Screenshot_2020-03-05 de tum in www1 artemis web rest(1)
Screenshot_2020-03-05 de tum in www1 artemis web rest(2)

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create and start modeling exercise with example assessment
4. Check that as an instructor you can update it.
5. Participate and submit a solution as a student.
6. Assess this exercise as a tutor.

### CODE REVIEW

**Comment by krusche** [comment]
Still a draft and not much progress since 17 days. You can reopen it, if the changes are finished!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java [hunk 1]
```diff
@@ -116,17 +116,18 @@ public ModelingExerciseResource(ModelingExerciseRepository modelingExerciseRepos
...
      *
      * @param modelingExercise the modelingExercise to update
+     * @param exerciseId the id of the exercise
      * @param notificationText the text shown to students
      * @return the ResponseEntity with status 200 (OK) and with body the updated modelingExercise, or with status 400 (Bad Request) if the modelingExercise is not valid, or with
      *         status 500 (Internal Server Error) if the modelingExercise couldn't be updated
      * @throws URISyntaxException if the Location URI syntax is incorrect
      */
-    @PutMapping("/modeling-exercises")
+    @PutMapping("/modeling-exercises/{exerciseId}")
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: Why is the entity id a path variable here? As far as I can see it's normally not included explicitly in the path for our PUT endpoints. Is there a specific reason why it's needed here?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java [hunk 2]
```diff
@@ -323,9 +323,9 @@ else if (!authCheckService.isAtLeastInstructorForExercise(exercise)) {
      * @param participationId the participationId for which to find the submission and data for the modeling editor
      * @return the ResponseEntity with the submission as body
      */
-    @GetMapping("/modeling-editor/{participationId}")
+    @GetMapping("/participations/{participationId}/latest-submission")
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java]: The naming seems very generic. Shouldn't `modeling` be in it somewhere since it only supports modeling exercises?

File: src/test/java/de/tum/in/www1/artemis/ModelingExerciseIntegrationTest.java [hunk 3]
```diff
@@ -117,15 +117,16 @@ public void testCreateModelingExercise_asInstructor() throws Exception {
     public void testUpdateModelingExercise_asInstructor() throws Exception {
         ModelingExercise modelingExercise = modelingExerciseUtilService.createModelingExercise(classExercise.getCourse().getId());
         // The PUT request basically forwards to POST in case the modeling exercise id is not yet set.
```
Comment (madwau) [inline src/test/java/de/tum/in/www1/artemis/ModelingExerciseIntegrationTest.java]: The comment is no longer relevant.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java (hunk 2 above)
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java]: True. Missed it somehow. I renamed it to `latest-modeling-submission`

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java (hunk 1 above)
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: The main reason is to conform to the standards defined for REST endpoints. `PUT` endpoints should always include the entity id. I did couple of changes to make it clear that we need it in the code. Now we also check if the modeling exercise exists in the database prior to storing update modeling exercise

Comment (madwau) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: Oh but then the behavior of the endpoint has been changed by the refactoring since it could be used before for both creating and updating whereas now it can only be used for updating. Did you check if all current usages are still valid with the new behavior?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java [hunk 4]
```diff
@@ -116,25 +111,28 @@ public ModelingExerciseResource(ModelingExerciseRepository modelingExerciseRepos
...
-            return createModelingExercise(modelingExercise);
-        }
 
         ResponseEntity<ModelingExercise> responseFailure = checkModelingExercise(modelingExercise);
         if (responseFailure != null) {
             return responseFailure;
         }
 
+        // Check that the modeling exercise id is correct
+        if (!modelingExerciseRepository.existsById(exerciseId) || !exerciseId.equals(modelingExercise.getId())) {
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: I think the second condition should have a different error message since that could be pretty confusing otherwise if the exercise actually exists but there is just an id mismatch.

Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: You are right I split it into two checks. First is done on the `ModelingExerciseService` and second one here. The second one now says `The updated modeling exercise doesn't match the given id` if failed.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java (hunk 1 above)
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: Yes, I checked that the creation and updating of the modeling exercise works properly. I also decided to change the update for the modeling submission. This endpoint is a bit more dangerous so I will still have to test it.

File: src/test/java/de/tum/in/www1/artemis/ModelingExerciseIntegrationTest.java (hunk 3 above)
Comment (maxr96) [reply src/test/java/de/tum/in/www1/artemis/ModelingExerciseIntegrationTest.java]: Good catch!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java (hunk 1 above)
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: **UPDATE**
Finished with the testing. Can you check it now?

**Comment by maxr96** [comment]
> @maxr96 please note: due to the recent critical bugs and problems in your changes, we need full test coverage of all changed files in the server and in the client, otherwise we need to decline this PR. Please have a look into the coverage reports and make sure all changed files are fully covered in the tests. If this is not the case, please add new tests to cover everything!
> 
> Please document this by uploading a screenshot of the coverage report to the PR.

I added test coverage reports for the files I changed. The lines that I changed in this PR are covered with the existing tests.

**Comment by krusche** [comment]
please fix the merge conflicts and make sure to further increase the test coverage, also the one for the changed client files!

**Comment by madwau** [comment]
@maxr96 Can you resolve the merge conflicts here?

File: src/main/java/de/tum/in/www1/artemis/service/ModelingExerciseService.java [hunk 5]
```diff
@@ -29,4 +29,10 @@ public ModelingExercise findOne(Long exerciseId) {
         log.debug("Request to get Modeling Exercise : {}", exerciseId);
         return modelingExerciseRepository.findById(exerciseId).orElseThrow(() -> new EntityNotFoundException("Exercise with id: \"" + exerciseId + "\" does not exist"));
     }
+
+    public void existsById(long exerciseId) {
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/service/ModelingExerciseService.java]: Applied

File: src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java [hunk 6]
```diff
@@ -386,4 +386,10 @@ public long countSubmissionsToAssessByCourseId(Long courseId) {
     public long countSubmissionsToAssessByExerciseId(Long exerciseId) {
         return modelingSubmissionRepository.countByExerciseIdSubmittedBeforeDueDate(exerciseId);
     }
+
+    public void existsById(long submissionId) {
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java]: Thank you for the suggestion. This definitely makes more sense

**Comment by maxr96** [comment]
Thank you for checking my PR @madwau and @ungaralex . I resolved the merge conflicts and applied the comments. I also increased the test coverage for `ModelingSubmissionResource` from 93% to 97% and `ModelingExerciseResource` from 93% to 96%. Can you check the PR again please?

**Comment by krusche** [comment]
@maxr96 this is still not 100%. Again, we only invest time in this PR, if you can make sure that really *ALL* functionality related to your changes is fully tested on the client and server!

**Comment by maxr96** [comment]
@krusche This line is not reachable as the results of participations are never initialized. That is why the code coverage of `ModelingAssessmentResource` is not 100%. However, for other two resources the coverage is now **exactly** 100%.
Screenshot_2020-03-05 ModelingAssessmentResource java

**Comment by maxr96** [comment]
> Hi Maksym, thanks for your contribution. I experienced the following unexpected behavior:
> 
>     * I am not able to create an example solution for a modeling exercise and get the following console error
>       (same with example submission)
>       image
>       Could you please fix this? Thanks.

Hi @HanyaElhashemy , thank you for checking my PR and reporting the issue. I checked that my changes did not cause this bug, as on the latest version of development branch I experience the same issue. However, I will investigate it and check how it can be fixed.

**Comment by maxr96** [comment]
**UPDATE**: I managed to solve the issue with modeling example solution and submission. Turns out that `ArtemisModelingExerciseManagementModule` was simply not imported into the `CourseManagementModule`. I also fixed back button routes for this components. Can you please chek my PR again @HanyaElhashemy ?

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts [hunk 7]
```diff
@@ -27,11 +27,11 @@ export class ModelingExerciseService {
             .pipe(map((res: EntityResponseType) => this.exerciseService.convertDateFromServer(res)));
     }
 
-    update(modelingExercise: ModelingExercise, req?: any): Observable<EntityResponseType> {
+    update(modelingExercise: ModelingExercise, exerciseId: number, req?: any): Observable<EntityResponseType> {
```
Comment (madwau) [inline src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts]: Isn't the `exerciseId` argument redundant here since it can be retrieved from `modelingExercise`?

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts [hunk 8]
```diff
@@ -24,10 +24,10 @@ export class ModelingSubmissionService {
             .map((res: EntityResponseType) => this.convertResponse(res));
     }
 
-    update(modelingSubmission: ModelingSubmission, exerciseId: number): Observable<EntityResponseType> {
+    update(modelingSubmission: ModelingSubmission, exerciseId: number, submissionId: number): Observable<EntityResponseType> {
```
Comment (madwau) [inline src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts]: Same here. Isn't the `submissionId` argument redundant here since it can be retrieved from `modelingSubmission`?

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts (hunk 7 above)
Comment (krusche) [reply src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts]: agreed

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java]: This makes our whole REST API inconsistent. I think we should revert this change!

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts (hunk 8 above)
Comment (krusche) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts]: I agree, let's revert this change!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java [hunk 9]
```diff
@@ -105,24 +105,29 @@ public ModelingSubmissionResource(ModelingSubmissionService modelingSubmissionSe
...
      *
      * @param exerciseId         the id of the exercise for which to init a participation
+     * @param submissionId       the id of the submission which needs to be updated
      * @param principal          the current user principal
      * @param modelingSubmission the modelingSubmission to update
      * @return the ResponseEntity with status 200 (OK) and with body the updated modelingSubmission, or with status 400 (Bad Request) if the modelingSubmission is not valid, or
      *         with status 500 (Internal Server Error) if the modelingSubmission couldn't be updated
      */
-    @PutMapping("/exercises/{exerciseId}/modeling-submissions")
+    @PutMapping("/exercises/{exerciseId}/modeling-submissions/{submissionId}")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java]: This makes our whole REST API inconsistent. I think we should revert this change!

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts (hunk 7 above)
Comment (maxr96) [reply src/main/webapp/app/exercises/modeling/manage/modeling-exercise.service.ts]: Ok, I reverted this change

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts (hunk 8 above)
Comment (maxr96) [reply src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts]: Ok, I reverted this change

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java (hunk 9 above)
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingSubmissionResource.java]: Reverted

**Comment by maxr96** [comment]
As requested I removed the ids in `ModelingSubmission` and `ModelingExercise` resources PUT endpoints
